### DISCUSSION TITLE
How do I add a gradient to the text?

### DISCUSSION THREAD

**Discussion by rebornfz** [libvips/libvips#2141] [Q&A] [answered]
Hello.
I want to add a gradient to the text, like this:
![下载](https://user-images.githubusercontent.com/27042311/110607548-466d2d80-81c6-11eb-80fd-e2f17b0f7054.png)

I have now generated such a picture
![91028907-0497e380-e5f6-11ea-9d23-b79715a1c72c](https://user-images.githubusercontent.com/27042311/110608247-065a7a80-81c7-11eb-8e53-326cf7a6d714.png)

How can i Make the color gradient?

**Comment by jcupitt** [comment]
Hello, make the gradient image then mask with the text. What sort of gradient do you need? Radial, linear, random, etc.?

**Reply by rebornfz** [reply]
linear gradient, how can i make the gradient image?

**Comment by jcupitt** [comment, marked as answer]
`grey` makes a greyscale ramp image with 0 on the left and 1 on the right:

https://libvips.github.io/libvips/API/8.9/libvips-create.html#vips-grey

So you can use that to blend between two colours like this:

```python
import pyvips

grey = pyvips.Image.grey(500, 500)
left = [255, 60, 140]
right = [0, 190, 220]
gradient = left * (1 - grey) + right * grey
```

![x](https://user-images.githubusercontent.com/580843/110622740-c29a4d80-8193-11eb-95f3-3901f451eb51.jpg)

**Reply by rebornfz** [reply]
Thank you, it works well. when i need a gradient image from top to bottom, i need to rotate image or some simple way？
Now, I'm using libvips to render some pictures to replace fabricjs, I have some problems when rendering text. I use vips_text() to make text image. I know vips_text can set the line spacing。but  I don't know how to adjust text spacing. Can you give me some advice？

**Reply by jcupitt** [reply]
Yes, eg.:

```python
grey = pyvips.Image.grey(500, 500).rot90()
```

Could you explain what you mean by text spacing? The space between words, or between letters?

**Reply by rebornfz** [reply]
The space between letters. like hello to h e l l o。

**Comment by jcupitt** [comment]
You can make any gradient function with `xyz`. For example:

```python
import pyvips

xyz = pyvips.Image.xyz(500, 500) - [250, 250]
circles = (1 + (5 * xyz.polar()[0]).sin()) / 2
left = [255, 60, 140]
right = [0, 190, 220]
gradient = left * (1 - circles) + right * circles
```

![x](https://user-images.githubusercontent.com/580843/110629303-ed889f80-819b-11eb-883c-94b96ffc9928.jpg)

Though obviously that's a bit slower.

**Reply by rebornfz** [reply]
It's great!
How can I adjust text spacing(the space between letters, like "hello" to "h e l l o" ), when I use the vips_text.(I use the C api)

**Reply by rebornfz** [reply]
I generally use Chinese, the word spacing maybe the space between letters

**Reply by jcupitt** [reply]
I replied re. text spacing ^^ up there, but github seems to truncate long reply threads, unfortunately. I reposted below.

**Comment by jcupitt** [comment]
For text spacing, Pango has a lot of attributes you can set, eg.:

```python
txt = pyvips.Image.text('<span letter_spacing="10000">Hello world!</span>', dpi=300)
```

![x](https://user-images.githubusercontent.com/580843/110643908-79a2c300-81ac-11eb-90d0-f770e4e1a9ef.png)

https://developer.gnome.org/pango/stable/pango-Markup.html

**Reply by rebornfz** [reply]
Thank you. I want every letter in a word has different style(every letter  maybe have different font, size, color.), like this
![下载 (1)](https://user-images.githubusercontent.com/27042311/110731998-46207100-825e-11eb-987a-9394db4eda50.png)
Do I need to render each letter individually? or some simple way?

**Reply by jcupitt** [reply]
Just set in the markup, eg.: `"<i>H</i>ell<b>o</b> <span font="sans 42">World</span>"`.

**Reply by rebornfz** [reply]
Thank you. I make it. I have changed the font, the size and the text spacing. But how can i get different color ? The vips_text() just make a band image. 
For example,
![google_logo](https://user-images.githubusercontent.com/27042311/110788849-1eef9100-82aa-11eb-9995-2ff9fd2d5965.png)
Do I  need to color each letter individually

**Comment by jcupitt** [comment]
Yes, `vips_text()` only renders black and white. You need to add any colour yourself.

We need to add a `colour` option.

**Reply by rebornfz** [reply]
Thank you. I am looking forward to this new feature
